File: packages/sdk-py/src/grantex/resources/_agents.py
from __future__ import annotations

from typing import Any, List, Optional

from .._http import HttpClient
from .._types import Agent, ListAgentsResponse


class AgentsClient:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def register(
        self,
        *,
        name: str,
        scopes: List[str],
        description: str = "",
        redirect_uris: List[str] | None = None,
        resource_servers: List[str] | None = None,
        public_jwk: dict[str, Any] | None = None,
    ) -> Agent:
        body: dict[str, Any] = {
            "name": name,
            "description": description,
            "scopes": scopes,
        }
        if redirect_uris is not None:
            body["redirectUris"] = redirect_uris
        if resource_servers is not None:
            body["resourceServers"] = resource_servers
        if public_jwk is not None:
            body["publicJwk"] = public_jwk
        data = self._http.post("/v1/agents", body)
        return Agent.from_dict(data)

    def get(self, agent_id: str) -> Agent:
        data = self._http.get(f"/v1/agents/{agent_id}")
        return Agent.from_dict(data)

    def list(self) -> ListAgentsResponse:
        data = self._http.get("/v1/agents")
        return ListAgentsResponse.from_dict(data)

    def update(
        self,
        agent_id: str,
        *,
        name: str | None = None,
        description: str | None = None,
        scopes: Optional[List[str]] = None,
        redirect_uris: Optional[List[str]] = None,
        resource_servers: Optional[List[str]] = None,
        public_jwk: dict[str, Any] | None = None,
    ) -> Agent:
        body: dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        if description is not None:
            body["description"] = description
        if scopes is not None:
            body["scopes"] = scopes
        if redirect_uris is not None:
            body["redirectUris"] = redirect_uris
        if resource_servers is not None:
            body["resourceServers"] = resource_servers
        if public_jwk is not None:
            body["publicJwk"] = public_jwk
        data = self._http.post(f"/v1/agents/{agent_id}", body)
        return Agent.from_dict(data)

    def delete(self, agent_id: str) -> None:
        self._http.delete(f"/v1/agents/{agent_id}")

File: packages/conformance/src/suites/agents.ts
import type { SuiteDefinition, SuiteContext, TestResult } from '../types.js';
import { test, skip, expectStatus, expectKeys, expectString, expectArray } from '../helpers.js';

export const agentsSuite: SuiteDefinition = {
  name: 'agents',
  description: 'Agent registration and management (CRUD)',
  optional: false,
  run: async (ctx: SuiteContext): Promise<TestResult[]> => {
    const results: TestResult[] = [];
    const sharedId = ctx.sharedAgent.agentId;
    let crudAgentId = '';

    // Try to create a fresh agent for CRUD tests
    const createRes = await ctx.http.post<{
      agentId: string;
      did: string;
      name: string;
      scopes: string[];
    }>('/v1/agents', {
      name: `conformance-crud-${Date.now()}`,
      scopes: ['read', 'write'],
    });

    if (createRes.status === 201) {
      crudAgentId = createRes.body.agentId;
    }

    results.push(
      crudAgentId
        ? await test('POST /v1/agents creates agent with agentId and did', '§10', async () => {
            expectKeys(createRes.body, ['agentId', 'did', 'name', 'scopes', 'status', 'createdAt']);
            expectString(createRes.body.agentId, 'agentId');
            expectString(createRes.body.did, 'did');
          })
        : skip('POST /v1/agents creates agent with agentId and did', '§10', 'Plan limit reached — cannot create test agent'),
    );

    results.push(
      await test('GET /v1/agents lists agents', '§10', async () => {
        const res = await ctx.http.get<{ agents: unknown[] }>('/v1/agents');
        expectStatus(res, 200);
        expectKeys(res.body, ['agents']);
        expectArray(res.body.agents, 'agents');
      }),
    );

    results.push(
      await test('GET /v1/agents/:id returns agent details', '§10', async () => {
        const id = crudAgentId || sharedId;
        const res = await ctx.http.get<{ agentId: string }>(`/v1/agents/${id}`);
        expectStatus(res, 200);
        expectKeys(res.body, ['agentId', 'did', 'name', 'scopes', 'status']);
      }),
    );

    results.push(
      await test('PATCH /v1/agents/:id updates agent', '§10', async () => {
        const id = crudAgentId || sharedId;
        const newName = `updated-${Date.now()}`;
        const res = await ctx.http.patch<{ agentId: string; name: string }>(
          `/v1/agents/${id}`,
          { name: newName },
        );
        expectStatus(res, 200);
        expectKeys(res.body, ['agentId', 'name']);
      }),
    );

    results.push(
      crudAgentId
        ? await test('DELETE /v1/agents/:id returns 204', '§10', async () => {
            const res = await ctx.http.delete(`/v1/agents/${crudAgentId}`);
            expectStatus(res, 204);
          })
        : skip('DELETE /v1/agents/:id returns 204', '§10', 'Plan limit reached — no test agent to delete'),
    );

    return results;
  },
};
